### PR TITLE
Adding resource types to the plugin specification.

### PR DESCRIPTION
**Pull request by albertshau**

### CODE REVIEW

File: config/defaults/automatortypes/chef-solo.json [hunk 1]
```diff
@@ -1,6 +1,17 @@
 {
     "description": "",
     "name": "chef-solo",
+    "resourceTypes": {
+        "cookbooks": {
+            "format": "archive"
+        },
+        "databags": {
```
Comment (wolf31o2) [inline config/defaults/automatortypes/chef-solo.json]: data_bags

File: server/src/main/java/com/continuuity/loom/admin/PluginType.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    super(name);
+    this.description = description == null ? "" : description;
+    this.parameters = parameters == null ? ImmutableMap.<ParameterType, ParametersSpecification>of() : parameters;
+    this.resourceTypes = resourceTypes == null ? ImmutableMap.<String, ResourceTypeSpecification>of() : resourceTypes;
+  }
+
+  /**
+   * Get the description of the automator type.
+   *
+   * @return Description of the automator type.
```
Comment (wolf31o2) [inline server/src/main/java/com/continuuity/loom/admin/PluginType.java]: not automator... plugin

File: server/src/test/java/com/continuuity/loom/Entities.java [hunk 3]
```diff
@@ -266,12 +287,28 @@
           ImmutableSet.<Set<String>>of(
             ImmutableSet.of("recipe")
           )
+        )),
+        ImmutableMap.<String, ResourceTypeSpecification>of(
+          "cookbooks", new ResourceTypeSpecification(ResourceTypeFormat.ARCHIVE),
+          "databags", new ResourceTypeSpecification(ResourceTypeFormat.ARCHIVE),
```
Comment (wolf31o2) [inline server/src/test/java/com/continuuity/loom/Entities.java]: data_bags

File: server/src/test/java/com/continuuity/loom/Entities.java [hunk 4]
```diff
@@ -266,12 +287,28 @@
...
+      );
     public static final String CHEF_STRING =
       "{\n" +
         "    \"name\": \"chef-solo\",\n" +
         "    \"description\": \"chef automator\",\n" +
+        "    \"resourceTypes\": {\n" +
+        "        \"cookbooks\": {\n" +
+        "            \"format\": \"archive\"\n" +
+        "        },\n" +
+        "        \"databags\": {\n" +
```
Comment (wolf31o2) [inline server/src/test/java/com/continuuity/loom/Entities.java]: data_bags

File: server/src/test/java/com/continuuity/loom/Entities.java [hunk 5]
```diff
@@ -309,8 +346,12 @@
           ImmutableSet.<Set<String>>of(
             ImmutableSet.of("manifest")
           )
+        )),
+        ImmutableMap.<String, ResourceTypeSpecification>of(
+          "modules", new ResourceTypeSpecification(ResourceTypeFormat.ARCHIVE),
+          "manifests", new ResourceTypeSpecification(ResourceTypeFormat.FILE)
```
Comment (wolf31o2) [inline server/src/test/java/com/continuuity/loom/Entities.java:352]: Puppet?

File: server/src/test/java/com/continuuity/loom/Entities.java [hunk 6]
```diff
@@ -76,12 +78,14 @@
...
-        ParametersSpecification.EMPTY_SPECIFICATION
-      ));
+        ParametersSpecification.EMPTY_SPECIFICATION),
+      ImmutableMap.<String, ResourceTypeSpecification>of()
+    );
     public static final String JOYENT_STRING =
       "{\n" +
         "    \"name\": \"joyent\",\n" +
         "    \"description\": \"joyent provider type\",\n" +
+        "    \"resourceTypes\": { },\n" +
```
Comment (wolf31o2) [inline server/src/test/java/com/continuuity/loom/Entities.java:88]: This is `resourceTypes` in the JSON?

File: server/src/main/java/com/continuuity/loom/admin/PluginType.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+
+/**
+ * Plugin type
+ */
+public abstract class PluginType extends NamedEntity {
```
Comment (poornachandra) [inline server/src/main/java/com/continuuity/loom/admin/PluginType.java]: Calling it AbstractPluginType will give this more meaning.

**Comment by poornachandra** [comment]
One minor comment from me, otherwise LGTM :+1:

File: config/defaults/automatortypes/chef-solo.json (hunk 1 above)
Comment (albertshau) [inline config/defaults/automatortypes/chef-solo.json]: done

File: server/src/main/java/com/continuuity/loom/admin/PluginType.java (hunk 2 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/admin/PluginType.java]: agh good catch

File: server/src/test/java/com/continuuity/loom/Entities.java (hunk 3 above)
Comment (albertshau) [inline server/src/test/java/com/continuuity/loom/Entities.java]: will do for consistency but doesn't really matter for unit tests

File: server/src/test/java/com/continuuity/loom/Entities.java (hunk 6 above)
Comment (albertshau) [inline server/src/test/java/com/continuuity/loom/Entities.java:88]: yep an example where there are none.

**Pull request merged**